src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp

```diff
@@ -0,0 +1,586 @@
+
+            watt.guid.guidPrefix = pdata->m_guid.guidPrefix;
+            watt.guid.entityId = c_EntityId_SPDPWriter;
+            watt.endpoint.persistence_guid = watt.guid;
+            watt.endpoint.unicastLocatorList = pdata->m_metatrafficUnicastLocatorList;
+            watt.endpoint.multicastLocatorList = pdata->m_metatrafficMulticastLocatorList;
+            watt.endpoint.reliabilityKind = RELIABLE;
+            watt.endpoint.durabilityKind = TRANSIENT;
+            watt.endpoint.topicKind = WITH_KEY;
+
+            // TODO remove the join when Reader and Writer match functions are updated
+            watt.endpoint.remoteLocatorList.push_back(pdata->m_metatrafficUnicastLocatorList);
+            watt.endpoint.remoteLocatorList.push_back(pdata->m_metatrafficMulticastLocatorList);
+
+            mp_PDPReader->matched_writer_remove(watt);
+            // rematch but discarding any previous state of the server
+            // because we know the server shutdown intencionally (sent a DATA(p[UD]))
+            mp_PDPReader->matched_writer_add(watt,false);
+        }
+
+        auxendp = endp;
+        auxendp &= DISC_BUILTIN_ENDPOINT_PARTICIPANT_DETECTOR;
+
+        if (auxendp != 0)
+        {
+            RemoteReaderAttributes ratt;
+
+            ratt.expectsInlineQos = false;
+            ratt.guid.guidPrefix = pdata->m_guid.guidPrefix;
+            ratt.guid.entityId = c_EntityId_SPDPReader;
+            ratt.endpoint.unicastLocatorList = pdata->m_metatrafficUnicastLocatorList;
+            ratt.endpoint.multicastLocatorList = pdata->m_metatrafficMulticastLocatorList;
+            ratt.endpoint.reliabilityKind = RELIABLE;
+            ratt.endpoint.durabilityKind = TRANSIENT_LOCAL;
+            ratt.endpoint.topicKind = WITH_KEY;
+
+            mp_PDPWriter->matched_reader_remove(ratt);
+            mp_PDPWriter->matched_reader_add(ratt);
+        }
+    }
+}
+
+bool PDPClient::all_servers_acknowledge_PDP()
+{
+    // check if already initialized
+    assert(mp_PDPWriterHistory && mp_PDPWriter);
+
+    // get a reference to client proxy data
+    CacheChange_t * pPD;
+    if (mp_PDPWriterHistory->get_min_change(&pPD))
+    {
+       return mp_PDPWriter->is_acked_by_all(pPD);
+    }
+    else
+    {
+        logError(RTPS_PDP, "ParticipantProxy data should have been added to client PDP history cache "
+            "by a previous call to announceParticipantState()");
+    }
+
+    return false;
+}
+
+bool PDPClient::is_all_servers_PDPdata_updated()
+{
+    // Assess all server DATA has been received
+    StatefulReader * pR = dynamic_cast<StatefulReader *>(mp_PDPReader);
+    assert(pR);
+    return pR->isInCleanState();
+}
+
+void PDPClient::announceParticipantState(
+    bool new_change,
+    bool dispose,
+    WriteParams& )
+{
+    WriteParams wp;
+    SampleIdentity local;
+    local.writer_guid(mp_PDPWriter->getGuid());
+    local.sequence_number(mp_PDPWriterHistory->next_sequence_number());
```

Comment:
`next_sequence_number` is non-thread-safe.

Reply:
Agreed!
I introduce a lock to enforce atomic access.

Reply:
Possible deadlock. Make sure always we lock in the same order: first the PDP mutex and then the PDP writer mutex.

Reply:
Basically we must get sure that we lock first on PDP mutex and then PDP Writer mutex (always in the same order) to prevent an AB BA deadlock.
		
		A - PDP mutex
		B - Writer mutex
		
		On a transport callback -> (B - A) PDPClient::announceParticipantState
		
		Who else calls PDPClient::announceParticipantState:
			- ~BuiltinProtocols on removal. That basically does (B - A) PDPClient::announceParticipantState.
			- BuiltinProtocols::initBuiltinProtocols. That basically does (B - A) PDPClient::announceParticipantState.
			- DSClientEvent thread. That basically does (B - A) PDPClient::announceParticipantState.
			- ResendParticipantProxyDataPeriod::event  That basically does (B - A) PDPClient::announceParticipantState.
			
		Thus no deadlock risk takes place.
		
		Study if the server is in the same situation. On a transport callback -> (B - A) PDPServer::announceParticipantState
		
		Who else calls PDPServer::announceParticipantState:
			- ~BuiltinProtocols on removal. That basically does (B - A) PDPServer::announceParticipantState.
			- BuiltinProtocols::initBuiltinProtocols. That basically does (B - A) PDPServer::announceParticipantState.
			- DServerEvent thread. That basically does (B - A) PDPServer::announceParticipantState.
			- ResendParticipantProxyDataPeriod::event  That basically does (B - A) PDPServer::announceParticipantState.
			
		Thus no deadlock risk takes place.